go-controller/pkg/ovn/ovn.go

```diff
@@ -537,8 +537,32 @@ func (oc *Controller) WatchPods() {
 		AddFunc: func(obj interface{}) {
 			pod := obj.(*kapi.Pod)
 			go oc.metricsRecorder.AddPod(pod.UID)
-			oc.initRetryAddPod(pod)
 			oc.checkAndSkipRetryPod(pod)
+			// in case ovnkube-master is restarted and gets all the add events with completed pods
+			if util.PodCompleted(pod) {
+				// pod is in completed state, remove it
+				klog.Infof("Detected completed pod: %s. Will remove.", getPodNamespacedName(pod))
+				oc.initRetryDelPod(pod)
+				oc.removeAddRetry(pod)
+				oc.logicalPortCache.remove(util.GetLogicalPortName(pod.Namespace, pod.Name))
+				retryEntry := oc.getPodRetryEntry(pod)
+				var portInfo *lpInfo
+				if retryEntry != nil {
+					// retryEntry shouldn't be nil since we usually add the pod to retryCache above
+					portInfo = retryEntry.needsDel
+				}
+				if err := oc.removePod(pod, portInfo); err != nil {
+					oc.recordPodEvent(err, pod)
+					klog.Errorf("Failed to delete completed pod %s, error: %v",
+						getPodNamespacedName(pod), err)
+					oc.unSkipRetryPod(pod)
+					return
+				}
+				oc.checkAndDeleteRetryPod(pod)
+				return
```

Comment:
I want to ask can we really get a pod add for a completed pod? :D 
I guess its more to handle a case where "foo" pod was in completed state where deletion was being retried, now we get a new pod with same name "foo" right? in this case we won't try out the new pod add even if the deletion was a success since we are returning at L562. Is that desired?

Reply:
I think we can when ovnkube master restarts and the informer sends us all add events